Improve documentation: Add rationale and better document `fromYAML` (#69)
dlang-community/configy: doc/rationale.md

# Configy rationale

The following explains the rationale versus various decisions made in Configy.

## Use cases, Internal vs External schema

Configy was originally designed with an application-first mindset: a developer should
design their config file based on a `struct` definition they chose, and would potentially
have a slightly different configuration file depending on the problem domain (for example,
the `@Key("name")` attribute).

While this is still the primary use-case for Configy, it is also very convenient to

> write a validator for schema that are outside of the application's control,

Reviewer: This is a little confusing to be because writing a validator for a schema seems orthogonal to being able to read schemas outside the application's control.

For validation, probably the best tool is a schema definition such as https://json-schema.org/.

Author: I have used it on a few occasions to quickly (as in, fast development time) validate that the config file passed is valid (I can give you links on Slack). I have been very pleased with the experience, but it might be a case of me only having ~Configy~ a hammer :)
We tried json-schema recently and it was very verbose, so we ended up using the library for that (but json-schema would have otherwise been my goto).

Reviewer: Ah, that's great. But I think that should be highlighted separately.

Validating a configuration file (and discarding the parsed values) seems to be one use case, along with just reading it (and validating as part of due course).

Author: > Validating a configuration file (and discarding the parsed values) seems to be one use case

But that any JSON / YAML parser can do, Configy is only about reading & validating. IMO it doesn't make sense to separate deserialization and validation, as the two are usually entangled (e.g. you need to validate anything that will affect parsing, such as the enumerated values in the `fromYAML` example).

Reviewer: First, I think it does make sense to separate them, as these are two different use cases. For example a validator could focus on providing good error messages and support additional validation semantics (i.e. validating domain-specific types), while a deserializer only cares about the data. Configy does both, but that's not true for all similar projects, so it's good to be aware of the distinction in the documentation.

Second... I'm not sure where this is going. If you say "we need X for validation" but then explain that X is actually used not just for validation but also deserialization I think that's confusing, because deserialization necessarily involves some degree of validation anyway.

Author: I think I might be overthinking this. Would the following work in your opinion ?
```suggestion
read and validate a schema that are outside of the application's control,
```

Reviewer: Yep!

(Also s/are/is/)

such as `dub.json`, `.gitlab-ci.yml` / GitHub actions file, or a replacement for
a JSON schema.

The first use case (when the schema is under the developer's control) is referred to
as an "Internal schema", while the second use case (when the schema is outside of the
developer's control) is referred to as an "External schema". While both are supported,
and Configy aims to have sane default that are likely to appear in external schemas,
it might sometimes be required to "drop down" to using hooks (`fromString`, `fromYAML`)

> to accurately control deserialization for external schema.

Reviewer: Whether the schema is internal or external seems to be a distraction here.

The facts are:

- Configy primarily supports an _opinionated subset_ of all possible YAML schemas.
- Application authors may want to read schemas outside of this subset, either because they're not under their control ("external") or they just disagree with Configy's opinionated decisions.

Author: That's a very good point. Edited accordingly.

## Data vs Identity

Configy being a DLang library adheres to one of the core tenet of the language,
which is to have a clear separation between data (stored in `struct`) and objects
with an identity (stored in `class`es). As configuration is data, Configy explicitly
does not support classes.

## The `struct` as the source of truth

When designing with internal schema in mind, there should be a top-level `struct`
that completely defines the program's configuration. Almost all of Configy's logic
will be driven by this `struct`. It is not possible to have fields that are ignored

> by Configy - configuration shouldn't be mixed with non-configuration.

Reviewer: I would also mention that we don't want to mix data with behaviour, so while struct methods are not technically forbidden, they are discouraged.

As much of the configuration properties as possible should be derived from
that `struct`. The main example of this design is the handling of default values:
if there is an initializer, it means there is a *sane* default value provided
by the developer, hence it needs not be a required field in the configuration.

Another opiniated design decision is that all fields are known, and required by default.
All fields being known means that, by default, Configy will error out if the document
contains fields that are not present in the `struct`. While there are various legitimate
use cases for unknown fields (such as forward-compatible configuration parsing and

> variable field names), being strict by default avoid the common pitfall of user typo,

Reviewer: ```suggestion
variable field names), being strict by default avoids the common pitfall of user typo,
```

which is a problem for every configuration file.

## External types

When designing a config format, even for internal schema, it might be useful to
have types that are external, such as library types (be it Phobos, Vibe.d, or another).
Those types are unlikely to have `fromYAML` support, and hence might need to be wrapped
in another project-controlled type to be present in the configuration struct. To avoid
needless wrapping, Configy supports the most common construction protocols: field-wise
constructor (the default constructor), string constructor, or `fromString`. Anything
else is likely to require a wrapping type that uses one of those methods, or `fromYAML`.

## Escape hatch

In order to support both external schema, and unforeseen internal schema use cases,
Configy provides a powerful escape hatch in the form of the `static` `fromYAML`

> method. In order to avoid loosing all the convenience of Configy once one "drops down"

Reviewer: ```suggestion
method. In order to avoid losing all the convenience of Configy once one "drops down"
```

to that level, the `ConfigParser` exposes a `parseAs` method which allows one to
recurse into the regular Configy parsing.

## Not fit for API or general purpose

The goal of Configy is to parse **configuration files**. For most applications, this
happens once per program run. For some, it may happens a few rare times (such as when

> the program is reload). As a result, and for the sake of simplicity, Configy does not

Reviewer: ```suggestion
the service is reloaded). As a result, and for the sake of simplicity, Configy does not
```
